engine/src/main/scala/cromwell/engine/workflow/lifecycle/deletion/DeleteWorkflowFilesActor.scala

```diff
@@ -1,44 +1,98 @@
 import cromwell.engine.workflow.lifecycle.deletion.DeleteWorkflowFilesActor._
+import cromwell.filesystems.gcs.batch.GcsBatchCommandBuilder
+import cromwell.services.metadata.MetadataService.PutMetadataAction
+import cromwell.services.metadata.impl.FileDeletionStatus
+import cromwell.services.metadata.impl.FileDeletionStatus.{Failed, InProgress, Succeeded}
+import cromwell.services.metadata.{MetadataEvent, MetadataKey, MetadataValue}
 import cromwell.util.GracefulShutdownHelper.ShutdownCommand
-import wom.graph.GraphNodePort.{GraphNodeOutputPort, OutputPort}
+import wom.graph.GraphNodePort.OutputPort
 import wom.types.WomSingleFileType
 import wom.values.{WomArray, WomSingleFile, WomValue}
 
+import scala.concurrent.ExecutionContext
 import scala.util.Try
 
 class DeleteWorkflowFilesActor(rootWorkflowId: RootWorkflowId,
                                workflowFinalOutputs: CallOutputs,
                                workflowAllOutputs: CallOutputs,
-                               pathBuilders: List[PathBuilder]) extends LoggingFSM[DeleteWorkflowFilesActorState, DeleteWorkflowFilesActorStateData] {
+                               pathBuilders: List[PathBuilder],
+                               serviceRegistryActor: ActorRef,
+                               ioActorRef: ActorRef) extends LoggingFSM[DeleteWorkflowFilesActorState, DeleteWorkflowFilesActorStateData] with IoClientHelper {
+
+  implicit val ec: ExecutionContext = context.dispatcher
 
   /*
   building a path for a string output such as 'Hello World' satisfies the conditions of a
   being a relative file path for a DefaultPath which we don't want, so remove DefaultPathBuilder
   (this is because we only want cloud backend paths like gs://, s3://, etc.)
    */
   val pathBuildersWithoutDefault = pathBuilders.filterNot(_ == DefaultPathBuilder)
+  val gcsCommandBuilder = GcsBatchCommandBuilder
+  val asyncIO = new AsyncIo(ioActorRef, gcsCommandBuilder)
 
   startWith(Pending, NoData)
 
   when(Pending) {
     case Event(StartWorkflowFilesDeletion, NoData) =>
       val intermediateOutputs = gatherIntermediateOutputFiles(workflowAllOutputs.outputs, workflowFinalOutputs.outputs)
-      if (intermediateOutputs.nonEmpty) goto(DeletingIntermediateFiles) using DeletingIntermediateFilesData(intermediateOutputs)
+      if (intermediateOutputs.nonEmpty) {
+        self ! DeleteFiles
+        goto(DeleteIntermediateFiles) using DeletingIntermediateFilesData(intermediateOutputs)
+      }
       else {
         log.info(s"Root workflow ${rootWorkflowId.id} does not have any intermediate output files to delete.")
         stopSelf()
       }
   }
 
+  when(DeleteIntermediateFiles) {
+    case Event(DeleteFiles, DeletingIntermediateFilesData(intermediateFiles)) =>
+      // update deletion status in metadata
+      val deletionInProgressEvent = metadataEventForDeletionStatus(InProgress)
+      serviceRegistryActor ! PutMetadataAction(deletionInProgressEvent)
+
```

Comment:
How bad would it be if the server crashed between lines 61 and 65? I know there's no general solution to this problem but it's worth reasoning out whether or not we'd get into a bad state this way.

Reply:
We may want to ask others, but I believe with the "replay" behavior of the Workflow Actor the various states within the metadata tables _should_ remain fine. Our general solution is to use a CRDT for WorkflowStatus, and everything else, including deletion status(?), is deemed "eventually consistent".

Reply:
It will restart the deletion process. So one might see an deletion status as `In Progress` twice in metadata.